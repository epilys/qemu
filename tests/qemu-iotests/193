#!/bin/bash
#
# Test block-insert-node & block-remove-node commands
#
# Copyright (C) 2017 Manos Pitsidianakis
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

# creator
owner="Manos Pitsidianakis"

seq=`basename $0`
echo "QA output created by $seq"

here=`pwd`
status=1	# failure is the default!

_cleanup()
{
    _cleanup_test_img
    rm -rf "$TEST_IMG".s
    rm -rf "$TEST_IMG".e

}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
. ./common.rc
. ./common.filter

_supported_fmt qcow2
_supported_proto file
_supported_os Linux

function do_run_qemu()
{
    echo Testing: "$@" | _filter_imgfmt
    $QEMU -nographic -qmp-pretty stdio -serial none "$@"
    echo
}

function run_qemu()
{
    do_run_qemu "$@" 2>&1 | _filter_testdir | _filter_qemu | _filter_qmp\
                          | _filter_qemu_io | _filter_generated_node_ids
}

_make_test_img 64M
test_throttle=$($QEMU_IMG --help|grep throttle)
[ "$test_throttle" = "" ] && _supported_fmt throttle

echo
echo "== inserting node =="

run_qemu <<EOF
{ "execute": "qmp_capabilities" }

{ "execute": "blockdev-add",
  "arguments": {
    "node-name": "disk",
    "driver": "$IMGFMT",
    "file": {
      "driver": "file",
      "filename": "$TEST_IMG"
    }
  }
}
{ "execute": "device_add",
   "arguments": { "driver": "virtio-blk", "drive": "disk",
                  "id": "virtio0" } }
{
    "execute": "blockdev-snapshot-sync",
    "arguments": {
        "node-name": "disk",
        "snapshot-file": "$TEST_IMG.s",
        "format": "$IMGFMT",
        "snapshot-node-name": "node-B"
    }
}
{ "execute": "blockdev-add",
  "arguments": {
    "node-name": "throttle0",
    "driver": "throttle",
    "file": "disk"  }
}

{ "execute": "block-insert-node",
    "arguments": {
    "parent": "node-B",
    "child": "disk",
    "node": "throttle0" }
}
{ "execute": "query-block" }
{ "execute": "query-named-block-nodes" }
{ "execute": "quit" }
EOF

echo
echo "== inserting and removing node =="

run_qemu <<EOF
{ "execute": "qmp_capabilities" }

{ "execute": "blockdev-add",
  "arguments": {
    "node-name": "disk",
    "driver": "$IMGFMT",
    "file": {
      "driver": "file",
      "filename": "$TEST_IMG"
    }
  }
}
{"execute": "blockdev-add",
"arguments": {
"node-name": "throttle0", "driver": "throttle", "file": "disk", "throttle-group": "foo", "limits" : { "iops-total": 100 } } }
{ "execute": "device_add",
   "arguments": { "driver": "virtio-blk", "drive": "throttle0",
                  "id": "virtio0" } }
{"execute": "blockdev-add",
"arguments": {
"node-name": "throttle1", "driver": "throttle", "file": "disk", "throttle-group": "bar", "limits" : { "iops-total": 200 } } }
{"execute": "block-insert-node",
	"arguments": {
	"parent": "throttle0",
	"child": "disk",
	"node": "throttle1"}
}
{ "execute": "query-block" }
{ "execute": "query-named-block-nodes" }
{"execute": "block-remove-node",
	"arguments": {
	"parent": "throttle0",
	"child": "disk",
	"node": "throttle1"}
}
{"execute": "blockdev-del",
    "arguments": { "node-name": "throttle1" }
}
{ "execute": "query-block" }
{ "execute": "query-named-block-nodes" }
{ "execute": "quit" }
EOF

echo
echo "== inserting node below BB =="

run_qemu -drive file=$TEST_IMG,format=$IMGFMT,if=none,id=disk,node-name=disk0 <<EOF
{ "execute": "qmp_capabilities" }
{ "execute": "blockdev-add",
  "arguments": {
    "node-name": "throttle0",
    "driver": "throttle",
    "file": "disk0"  }
}
{ "execute": "block-insert-node",
    "arguments": {
    "parent": "disk",
    "child": "disk0",
    "node": "throttle0" }
}
{ "execute": "query-block" }
{ "execute": "query-named-block-nodes" }
{"execute": "block-remove-node",
	"arguments": {
	"parent": "disk",
	"child": "disk0",
	"node": "throttle0"}
}
{"execute": "blockdev-del",
    "arguments": { "node-name": "throttle0" }
}
{ "execute": "query-block" }
{ "execute": "query-named-block-nodes" }
{ "execute": "quit" }
EOF

echo
echo "== try insert with active block job =="

# this should error because of the hidden mirror filter on top
run_qemu -drive file=$TEST_IMG,format=$IMGFMT,if=none,id=disk,node-name=disk0 <<EOF
{ "execute": "qmp_capabilities" }
{
    "execute": "blockdev-snapshot-sync",
    "arguments": {
        "node-name": "disk0",
        "snapshot-file": "$TEST_IMG.s",
        "format": "$IMGFMT",
        "snapshot-node-name": "overlay0"
    }
}
{
    "execute": "drive-mirror",
    "arguments": {
        "device": "overlay0",
        "job-id": "job0",
        "target": "$TEST_IMG.e",
        "sync": "full"
    }
}
{
    "execute": "query-block-jobs",
    "arguments": {}
}
{ "execute": "blockdev-add",
  "arguments": {
    "node-name": "throttle0",
    "driver": "throttle",
    "file": "overlay0"  }
}
{ "execute": "block-insert-node",
    "arguments": {
    "parent": "disk",
    "child": "overlay0",
    "node": "throttle0" }
}
{ "execute": "query-block" }
{"execute": "blockdev-del",
    "arguments": { "node-name": "throttle0" }
}
{ "execute": "quit" }
EOF

echo
# success, all done
echo "*** done"
rm -f $seq.full
status=0
