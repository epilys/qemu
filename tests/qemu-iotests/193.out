QA output created by 193
Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=67108864

== inserting node ==
Testing:
{
    QMP_VERSION
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
Formatting 'TEST_DIR/t.qcow2.s', fmt=qcow2 size=67108864 backing_file=TEST_DIR/t.qcow2 backing_fmt=qcow2 cluster_size=65536 lazy_refcounts=off refcount_bits=16
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "device": "",
            "locked": false,
            "removable": false,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "backing-image": {
                        "virtual-size": 67108864,
                        "filename": "TEST_DIR/t.qcow2",
                        "format": "throttle",
                        "actual-size": 200704
                    },
                    "backing-filename-format": "throttle",
                    "virtual-size": 67108864,
                    "filename": "TEST_DIR/t.qcow2.s",
                    "cluster-size": 65536,
                    "format": "qcow2",
                    "actual-size": 200704,
                    "format-specific": {
                        "type": "qcow2",
                        "data": {
                            "compat": "1.1",
                            "lazy-refcounts": false,
                            "refcount-bits": 16,
                            "corrupt": false
                        }
                    },
                    "full-backing-filename": "TEST_DIR/t.qcow2",
                    "backing-filename": "TEST_DIR/t.qcow2",
                    "dirty-flag": false
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "node-B",
                "backing_file_depth": 1,
                "drv": "qcow2",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "backing_file": "TEST_DIR/t.qcow2",
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "TEST_DIR/t.qcow2.s",
                "encryption_key_missing": false
            },
            "qdev": "/machine/peripheral/virtio0/virtio-backend",
            "type": "unknown"
        }
    ]
}
{
    "return": [
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "format": "throttle",
                "actual-size": 200704
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "throttle0",
            "backing_file_depth": 0,
            "drv": "throttle",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "backing-image": {
                    "virtual-size": 67108864,
                    "filename": "TEST_DIR/t.qcow2",
                    "format": "throttle",
                    "actual-size": 200704
                },
                "backing-filename-format": "throttle",
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2.s",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "full-backing-filename": "TEST_DIR/t.qcow2",
                "backing-filename": "TEST_DIR/t.qcow2",
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "node-B",
            "backing_file_depth": 1,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "backing_file": "TEST_DIR/t.qcow2",
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2.s",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2.s",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2.s",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": true,
            "node-name": "disk",
            "backing_file_depth": 0,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": true,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        }
    ]
}
{
    "return": {
    }
}
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "SHUTDOWN",
    "data": {
        "guest": false
    }
}


== inserting and removing node ==
Testing:
{
    QMP_VERSION
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "device": "",
            "locked": false,
            "removable": false,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "virtual-size": 67108864,
                    "filename": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}}",
                    "format": "throttle",
                    "actual-size": 200704
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "throttle0",
                "backing_file_depth": 0,
                "drv": "throttle",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}}",
                "encryption_key_missing": false
            },
            "qdev": "/machine/peripheral/virtio0/virtio-backend",
            "type": "unknown"
        }
    ]
}
{
    "return": [
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "json:{\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
                "format": "throttle",
                "actual-size": 200704
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "throttle1",
            "backing_file_depth": 0,
            "drv": "throttle",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "json:{\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}}",
                "format": "throttle",
                "actual-size": 200704
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "throttle0",
            "backing_file_depth": 0,
            "drv": "throttle",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"throttle-group\": \"bar\", \"driver\": \"throttle\", \"limits.iops-total\": 200, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}}",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "disk",
            "backing_file_depth": 0,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        }
    ]
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "device": "",
            "locked": false,
            "removable": false,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "virtual-size": 67108864,
                    "filename": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
                    "format": "throttle",
                    "actual-size": 200704
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "throttle0",
                "backing_file_depth": 0,
                "drv": "throttle",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
                "encryption_key_missing": false
            },
            "qdev": "/machine/peripheral/virtio0/virtio-backend",
            "type": "unknown"
        }
    ]
}
{
    "return": [
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
                "format": "throttle",
                "actual-size": 200704
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "throttle0",
            "backing_file_depth": 0,
            "drv": "throttle",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "json:{\"throttle-group\": \"foo\", \"driver\": \"throttle\", \"limits.iops-total\": 100, \"file\": {\"driver\": \"qcow2\", \"file\": {\"driver\": \"file\", \"filename\": \"TEST_DIR/t.qcow2\"}}}",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "disk",
            "backing_file_depth": 0,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        }
    ]
}
{
    "return": {
    }
}
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "SHUTDOWN",
    "data": {
        "guest": false
    }
}


== inserting node below BB ==
Testing: -drive file=TEST_DIR/t.IMGFMT,format=IMGFMT,if=none,id=disk,node-name=disk0
{
    QMP_VERSION
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "device": "disk",
            "locked": false,
            "removable": true,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "virtual-size": 67108864,
                    "filename": "TEST_DIR/t.qcow2",
                    "format": "throttle",
                    "actual-size": 200704
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "throttle0",
                "backing_file_depth": 0,
                "drv": "throttle",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "TEST_DIR/t.qcow2",
                "encryption_key_missing": false
            },
            "type": "unknown"
        }
    ]
}
{
    "return": [
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "format": "throttle",
                "actual-size": 200704
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "throttle0",
            "backing_file_depth": 0,
            "drv": "throttle",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "disk0",
            "backing_file_depth": 0,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        }
    ]
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "device": "disk",
            "locked": false,
            "removable": true,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "virtual-size": 67108864,
                    "filename": "TEST_DIR/t.qcow2",
                    "cluster-size": 65536,
                    "format": "qcow2",
                    "actual-size": 200704,
                    "format-specific": {
                        "type": "qcow2",
                        "data": {
                            "compat": "1.1",
                            "lazy-refcounts": false,
                            "refcount-bits": 16,
                            "corrupt": false
                        }
                    },
                    "dirty-flag": false
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "disk0",
                "backing_file_depth": 0,
                "drv": "qcow2",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "TEST_DIR/t.qcow2",
                "encryption_key_missing": false
            },
            "type": "unknown"
        }
    ]
}
{
    "return": [
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 67108864,
                "filename": "TEST_DIR/t.qcow2",
                "cluster-size": 65536,
                "format": "qcow2",
                "actual-size": 200704,
                "format-specific": {
                    "type": "qcow2",
                    "data": {
                        "compat": "1.1",
                        "lazy-refcounts": false,
                        "refcount-bits": 16,
                        "corrupt": false
                    }
                },
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "disk0",
            "backing_file_depth": 0,
            "drv": "qcow2",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        },
        {
            "iops_rd": 0,
            "detect_zeroes": "off",
            "image": {
                "virtual-size": 197120,
                "filename": "TEST_DIR/t.qcow2",
                "format": "file",
                "actual-size": 200704,
                "dirty-flag": false
            },
            "iops_wr": 0,
            "ro": false,
            "node-name": "NODE_NAME",
            "backing_file_depth": 0,
            "drv": "file",
            "iops": 0,
            "bps_wr": 0,
            "write_threshold": 0,
            "encrypted": false,
            "bps": 0,
            "bps_rd": 0,
            "cache": {
                "no-flush": false,
                "direct": false,
                "writeback": true
            },
            "file": "TEST_DIR/t.qcow2",
            "encryption_key_missing": false
        }
    ]
}
{
    "return": {
    }
}
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "SHUTDOWN",
    "data": {
        "guest": false
    }
}


== try insert with active block job ==
Testing: -drive file=TEST_DIR/t.IMGFMT,format=IMGFMT,if=none,id=disk,node-name=disk0
{
    QMP_VERSION
}
{
    "return": {
    }
}
Formatting 'TEST_DIR/t.qcow2.s', fmt=qcow2 size=67108864 backing_file=TEST_DIR/t.qcow2 backing_fmt=qcow2 cluster_size=65536 lazy_refcounts=off refcount_bits=16
{
    "return": {
    }
}
Formatting 'TEST_DIR/t.qcow2.e', fmt=qcow2 size=67108864 cluster_size=65536 lazy_refcounts=off refcount_bits=16
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "BLOCK_JOB_READY",
    "data": {
        "device": "job0",
        "len": 0,
        "offset": 0,
        "speed": 0,
        "type": "mirror"
    }
}
{
    "return": {
    }
}
{
    "return": [
        {
            "io-status": "ok",
            "device": "job0",
            "busy": false,
            "len": 0,
            "offset": 0,
            "paused": false,
            "speed": 0,
            "ready": true,
            "type": "mirror"
        }
    ]
}
{
    "return": {
    }
}
{
    "error": {
        "class": "GenericError",
        "desc": "'overlay0' not a child of device 'disk'"
    }
}
{
    "return": [
        {
            "device": "disk",
            "locked": false,
            "removable": true,
            "inserted": {
                "iops_rd": 0,
                "detect_zeroes": "off",
                "image": {
                    "backing-image": {
                        "virtual-size": 67108864,
                        "filename": "TEST_DIR/t.qcow2",
                        "cluster-size": 65536,
                        "format": "qcow2",
                        "actual-size": 200704,
                        "format-specific": {
                            "type": "qcow2",
                            "data": {
                                "compat": "1.1",
                                "lazy-refcounts": false,
                                "refcount-bits": 16,
                                "corrupt": false
                            }
                        },
                        "dirty-flag": false
                    },
                    "backing-filename-format": "qcow2",
                    "virtual-size": 67108864,
                    "filename": "TEST_DIR/t.qcow2.s",
                    "cluster-size": 65536,
                    "format": "qcow2",
                    "actual-size": 200704,
                    "format-specific": {
                        "type": "qcow2",
                        "data": {
                            "compat": "1.1",
                            "lazy-refcounts": false,
                            "refcount-bits": 16,
                            "corrupt": false
                        }
                    },
                    "full-backing-filename": "TEST_DIR/t.qcow2",
                    "backing-filename": "TEST_DIR/t.qcow2",
                    "dirty-flag": false
                },
                "iops_wr": 0,
                "ro": false,
                "node-name": "overlay0",
                "backing_file_depth": 1,
                "drv": "qcow2",
                "iops": 0,
                "bps_wr": 0,
                "write_threshold": 0,
                "backing_file": "TEST_DIR/t.qcow2",
                "encrypted": false,
                "bps": 0,
                "bps_rd": 0,
                "cache": {
                    "no-flush": false,
                    "direct": false,
                    "writeback": true
                },
                "file": "TEST_DIR/t.qcow2.s",
                "encryption_key_missing": false
            },
            "dirty-bitmaps": [
                {
                    "status": "active",
                    "granularity": 65536,
                    "count": 0
                }
            ],
            "type": "unknown"
        }
    ]
}
{
    "return": {
    }
}
{
    "return": {
    }
}
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "SHUTDOWN",
    "data": {
        "guest": false
    }
}
{
    "timestamp": {
        "seconds":  TIMESTAMP,
        "microseconds":  TIMESTAMP
    },
    "event": "BLOCK_JOB_COMPLETED",
    "data": {
        "device": "job0",
        "len": 0,
        "offset": 0,
        "speed": 0,
        "type": "mirror"
    }
}


*** done
